docs/architecture/adr-047-consensus-fees.md

```diff
@@ -0,0 +1,181 @@
+
+We can adopt EIP-1559's approach to adjust higher tier's gas prices based on block load.
+
+### Implementation
+
+We can make each tier's gas price strategy fully configurable in protocol parameters, while providing a sensible default one.
+
+Pseudocode in python-like syntax:
+
+```python
+interface TieredTx:
+  def tier(self) -> int:
+    pass
+
+def tx_tier(tx):
+    if isinstance(tx, TieredTx):
+      return tx.tier()
+    else:
+      # default tier for custom transactions
+      return 0
+
+class TierParams:
+  'gas price strategy parameters of one tier'
+  priority: int           # priority in tendermint mempool
+  initial_gas_price: Coin
+  parent_gas_target: int
+  change_speed: Decimal   # 0 means don't adjust for block load.
+
+class Params:
+    'protocol parameters'
+    tiers: List[TierParams]
+
+class State:
+    'consensus state'
+    # total gas used in last block, None when it's the first block
+    parent_gas_used: Optional[int]
+    # gas prices of last block for all tiers
+    gas_prices: List[Coin]
+
+def begin_block():
+    'Adjust gas prices'
+    for i, tier in enumerate(Params.tiers):
+        if State.parent_gas_used is None:
+            # initialized gas price for the first block
+	          State.gas_prices[i] = tier.initial_gas_price
+        else:
+            gas_used_change_rate = (State.parent_gas_used - tier.parent_gas_target) // tier.parent_gas_target
+            State.gas_prices[i] *= 1 + gas_used_change_rate * tier.change_speed
+
+def mempoolFeeTxHandler_checkTx(ctx, tx):
+    # the minimal-gas-price configured by validator, zero in deliver_tx context
+    validator_price = ctx.MinGasPrice()
+    consensus_price = State.gas_prices[tx_tier(tx)]
+    min_price = max(validator_price, consensus_price)
+
+    if tx.gas_price() < min_price:
+        return 'insufficient fees'
+    return next_CheckTx(ctx, tx)
+
+def txPriorityHandler_checkTx(ctx, tx):
+    res, err := next_CheckTx(ctx, tx)
+    # pass priority to tendermint
+    res.Priority = Params.tiers[tx_tier(tx)].priority
+    return res, err
+
+def end_block():
+    'Update block gas used'
+    State.parent_gas_used = block_gas_meter.consumed()
+```
+
+### Dos attack protection
+
+To fully saturate the blocks and prevent other transactions from executing, attacker need to use transactions of highest tier, the cost would be significantly higher than the default tier.
+
+If attacker spam with lower tier transactions, user can mitigate by sending higher tier transactions.
+
+## Consequences
+
+### Backwards Compatibility
+
+- New protocol parameters.
+- New consensus states.
+- New optional field in transaction body.
+
+### Positive
+
+- The default tier keeps the same predictable gas price experience for client.
+- No priority conflict with custom priority based on transaction types, since this proposal only occupy three priority levels.
+
+### Negative
+
```

Comment:
this is perhaps the case for all dynamic fee systems, but one negative or neutral consequence will be that Tendermint's `create-empty-blocks = false` option won't work, because the state / app hash will keep changing at every block

Reply:
I think `create-empty-blocks = false` means the validator only proposal non-empty blocks, so it should still work?

Reply:
When `create-empty-blocks = false`, the validator doesn't propose new blocks when there's no tx, the only consequence is the gas price is not adjusted in the meantime?